### DISCUSSION TITLE
FastDDS 2.14.3 Dynamic Types with Type Object not working

### DISCUSSION THREAD

**Discussion by arnabgho81** [eProsima/Fast-DDS#5171] [Q&A]
Hi, I am using Fast DDS version 2.14.3 and fastDDSGen 3.0.1. I am able to generate the type Object from an Idl but when i try to create the dynamic type or the dynamic data from a serialized payload I face core dump.

Here is a sample Idl defination

module sample_cluster {

enum SampleClass
{
  UNKNOWN,
  A,
  B,
  C,
  D
  LOCAL
};

struct SampleCluster 
{
  double stamps_s; 
  boolean flag1; 
  unsigned short code;
  boolean flag2;
  SampleClass sample_class;
};

};

The FastDDSGen was able to generate the typeObject.

Then i use the code listed below:

    // type_ is the PubSubtype generated from the FastDDSGen 
    std::string type_name = type_->getName();
    std::cout << "Create Dynamic Type ::" << type_name << std::endl;
    
    // Use data type name, type identifier and type object to build the dynamic type
    const eprosima::fastrtps::types::TypeIdentifier* type_identifier = 
                      eprosima::fastrtps::types::TypeObjectFactory::get_instance()->get_type_identifier_trying_complete(type_name);
    const eprosima::fastrtps::types::TypeObject* type_object = 
                    eprosima::fastrtps::types::TypeObjectFactory::get_instance()->get_type_object(type_identifier);
    eprosima::fastrtps::types::DynamicType_ptr dynamic_type_ =
        eprosima::fastrtps::types::TypeObjectFactory::get_instance()->build_dynamic_type(type_name, type_identifier, type_object);

    eprosima::fastdds::dds::TypeSupport type(new eprosima::fastrtps::types::DynamicPubSubType(dynamic_type_));    
  
    std::cout << "Create Dynamic Data " << std::endl;

    //Create a new DynamicData to read the sample
    eprosima::fastrtps::types::DynamicData_ptr new_dynamic_data;
    new_dynamic_data = eprosima::fastrtps::types::DynamicDataFactory::get_instance()->create_data(dynamic_type_);

    std::cout << "Deserialize Data " << std::endl;

    type->deserialize(&serializedPayload, (void*)(new_dynamic_data.get()));

    std::cout << "Message received: " << std::endl;
    eprosima::fastrtps::types::DynamicDataHelper::print(new_dynamic_data);
    std::cout << "-----------------------------------------------------" << std::endl;   

I was able to verify that the serialized_payload is being deserialized using the PubSubtype.deserialize method.

**Comment by arnabgho81** [comment]
My query is similar to #5062

**Comment by JesusPoderoso** [comment]
Hi @arnabgho81, thanks for using Fast DDS.
This is a [known issue](https://github.com/eProsima/Fast-DDS/issues/3296) from Fast DDS 2.3.x and onwards (remaining 2.x versions). Dynamic types (XTypes 1.3) have been refactored in `master` branch and will be released soon as part of the API breaks for Fast DDS v3.0.0.
Could you check if the issue persists using `master` branch?
The [XTypes example](https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/xtypes) and the [migration guide](https://github.com/eProsima/Fast-DDS/blob/master/UPGRADING.md) may help you to move to v3.0.0.

**Reply by arnabgho81** [reply]
Which C++ compiler version should I use to compile the project

**Reply by JesusPoderoso** [reply]
Hi  @arnabgho81, they might be different compiler versions depending on the Fast DDS version you are using, 
Refer to the [build system dependencies](https://fast-dds.docs.eprosima.com/en/latest/notes/versions.html#build-system-dependencies) documentation section.

**Reply by arnabgho81** [reply]
Now that I have migrated to Fast DDS v3.0.0 the dynamic type is working for the idl.
![DynamicTypes_sample](https://github.com/user-attachments/assets/7e5f76ce-10ba-4de4-95f4-6dee9f52de9e)

I have related query, when I retrieve the TypeObject via ITypeObjectRegistry::get_type_object() will it be containing all the dependent typeobjects, which means the type object is self contained or do I need to get the dependent typeObject seperately.

**Reply by arnabgho81** [reply]
If i serialize the type object and then deserialize the type object, will the deserialized typeobject have all the dependent types to be used to create dynamic data.

**Reply by arnabgho81** [reply]
I am getting Memory Fault core dump at line marked with red.
![Dyanmic_error](https://github.com/user-attachments/assets/f70de56f-3f4e-450c-96fe-016a3be8691f)

**Reply by arnabgho81** [reply]
I found the bug in my code. It is working now.
![Dynamic_type_fix](https://github.com/user-attachments/assets/0b00fcda-754f-4f85-a983-0752efa64529)
 and dynamic type is working as expected.
